llvm/docs/LangRef.rst

```diff
@@ -650,48 +650,127 @@ literal types are uniqued in recent versions of LLVM.
+difference compared to integral pointers is that ``inttoptr`` instructions
+should not be inserted by passes as they may not be able to create a valid
+pointer. This property also means that ``inttoptr(ptrtoint(x))`` cannot be
+folded to ``x`` as the ``ptrtoint`` operation may destroy the necessary metadata
+to reconstruct the pointer.
+
+Unlike "unstable" pointers, the bit-wise representation is stable and
+``ptrtoint(x)`` always yields a deterministic value.
+This means transformation passes are still permitted to insert new ``ptrtoint``
+instructions.
+However, it is important to note that ``ptrtoint`` may not yield the same value
+as storing the pointer via memory and reading it back as an integer, even if the
+bitwidth of the two types matches (since ptrtoint could involve some form of
+arithmetic or strip parts of the non-integral pointer representation).
+
+Non-integral pointers with external state
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A special case of non-integral pointers is ones that include external state
+(such as implicit bounds information or a type tag) with a target-defined size.
+An example of such a type is a CHERI capability, where there is an additional
+validity bit that is part of all pointer-typed registers, but is located in
+memory at an implementation-defined address separate from the pointer itself.
+Another example would be a fat-pointer scheme where pointers remain plain
+integers, but the associated bounds are stored in an out-of-band table.
+
+When a ``store ptr addrspace(N) %p, ptr @dst`` of such a non-integral pointers
+is performed, the external metadata is also stored to another location.
+Similarly, a ``%val = load ptr addrspace(N), ptr @dst`` will fetch the
+external metadata and make it available for all uses of ``%val``.
+Notionally, these additional bits are part of the pointer, but since loads and
+stores only operate on the "inline" bits of the pointer and the additional
+bits are not explicitly exposed, they are not included in the size specified in
+the :ref:`datalayout string<langref_datalayout>`.
+
+When a pointer type has external state, all roundtrips via memory must
+be performed as loads and stores of the correct type since stores of other
+types may not propagate the external data.
+Therefore it is not legal to convert a load/store of a non-integral pointer
+type with external state to a load/store of an integer type with same
+bitwidth, as that drops the additional state.
+However, it can be assumed that appropriately-aligned ``llvm.memcpy`` and
+``llvm.memmove`` intrinsics will preserve the external metadata.
+This is essential to allow frontends to efficiently emit of copies of
+structures containing such pointers, since expanding all these copies as
+individual loads and stores would significantly inhibit optimizations.
+To ensure that this results in valid code, affected backends must lower
+these intrinsics in a way that propagates external state.
+
```

Comment:
Can we clarify "appropriately-aligned" here?

Reply:
I should probably drop "appropriately-aligned" - this relates to the fact that CHERI capabilities can only be loaded from/stored to naturally aligned memory locations since there is only one tag bit per capability-size location.